diff --git a/firmware/3D-renderer/Makefile b/firmware/3D-renderer/Makefile
index 530a44f..365cf53 100644
--- a/firmware/3D-renderer/Makefile
+++ b/firmware/3D-renderer/Makefile
@@ -26,7 +26,7 @@ SIMULATOR=$(LOCAL_TOOL_DIR)/simulator/simulator
 VERILATOR=../../rtl/obj_dir/Vverilator_tb
 MKBMP=$(LOCAL_TOOL_DIR)/mkbmp/mkbmp
 CFLAGS=-g -Wall -W -O3 -fno-rtti -std=c++11
-BASE_ADDRESS=0
+BASE_ADDRESS=0x10140000
 
 SRCS=main.cpp \
 	Rasterizer.cpp \
diff --git a/firmware/3D-renderer/main.cpp b/firmware/3D-renderer/main.cpp
index 9199e40..aa06e2b 100644
--- a/firmware/3D-renderer/main.cpp
+++ b/firmware/3D-renderer/main.cpp
@@ -53,7 +53,7 @@
 
 using namespace render;
 
-const int kFbWidth = 512;
+const int kFbWidth = 640;
 const int kFbHeight = 512;	// Round up to 64 pixel boundary
 
 const int kTilesPerRow = kFbWidth / kTileSize;
@@ -64,7 +64,7 @@ runtime::Barrier gInitBarrier;
 volatile int gNextTileIndex = 0;
 float *gVertexParams;
 render::Surface gZBuffer(0, kFbWidth, kFbHeight);
-render::Surface gColorBuffer(0x100000, kFbWidth, kFbHeight);
+render::Surface gColorBuffer(0x10000000, kFbWidth, kFbHeight);
 #if DRAW_CUBE
 	render::Surface texture((unsigned int) kBrickTexture, 128, 128);
 #endif
@@ -410,6 +410,6 @@ int main()
 		
 		gPixelBarrier.wait();
 	}
-	
+  
 	return 0;
 }
diff --git a/firmware/3D-renderer/start.s b/firmware/3D-renderer/start.s
index aabefb9..5406640 100644
--- a/firmware/3D-renderer/start.s
+++ b/firmware/3D-renderer/start.s
@@ -59,6 +59,6 @@ skip_init:			call main
 					setcr s0, 29 ; Stop thread, mostly for simulation
 done:				goto done
 
-stacks_base:		.long 0x100000
+stacks_base:		.long 0x10340000
 init_array_start:	.long __init_array_start
 init_array_end:		.long __init_array_end
diff --git a/firmware/3D-renderer/utils.cpp b/firmware/3D-renderer/utils.cpp
index 2388cee..3f0b4f2 100644
--- a/firmware/3D-renderer/utils.cpp
+++ b/firmware/3D-renderer/utils.cpp
@@ -25,7 +25,7 @@ extern "C"  {
 }
 
 void *__dso_handle;
-static volatile unsigned int gNextAlloc = 0x240000;
+static volatile unsigned int gNextAlloc = 0x10340000;
 
 void memcpy(void *dest, const void *src, unsigned int length)
 {
